vowpalwabbit/core/src/reductions/baseline_challenger_cb.cc

```diff
@@ -215,15 +220,27 @@ VW::LEARNER::base_learner* VW::reductions::baseline_challenger_cb_setup(VW::setu
 
   if (!options.was_supplied("cb_adf")) { THROW("cb_challenger requires cb_explore_adf or cb_adf"); }
 
-  bool emit_metrics = options.was_supplied("extra_metrics");
-  auto data = VW::make_unique<baseline_challenger_data>(emit_metrics, alpha, tau);
+  baseline_challenger_cb_opts->emit_metrics = options.was_supplied("extra_metrics");
```

Comment:
If this reduction is determining whether to emit metrics or not based on the options then it should be part of its option description

Reply:
isn't metrics one of those options that is universal across reductions?

Reply:
Yeah, epsilon is too and is part of the option group so we aren't consistent. I don't feel super confident about these cross-reduction global options because right now they apply to all reductions. However, if the concept of option namespacing is brought in then the global/cross reduction options no longer work as before

Reply:
epsilon is reduction-specific it is exploration only, this is more close to `random_seed`

Reply:
epsilon is still defined in 11 reductions, I think at that point its no longer all that specific

Reply:
you can't say that epsilon is universal though even though 11 exploration reductions use it, where as metrics can be used by every single reduction if they wish to do so

if we are going to make the change to move the universal reduction options into every reduction I think it should be done in a separate PR

Reply:
Yeah it is out of scope of this and I am not saying we even should do it. I'm mostly pointing out that unless we put an option into a reductions option description there is no way for a user to know that the reduction can use that option

Reply:
that is true

Reply:
Probably out of scope for this PR but wanted to mention here...

Some options seem to be vw_driver options... for example, threading, logging, metrics, prediction_file, input_file etc.  And some that are vw_ml (reduction_lib?) options, for example epsilon.  

Jacob mentioned possible support for vw_ml options, where individual reductions can have their own epsilon values, if specified, without any explicit code from the reduction author.

Reply:
I think with `extra_metrics` specifically it makes sense to parse the option elsewhere globally and add it to the all/logger object. I can open this in a separate PR, thoughts on this?

Reply:
Totally agree on clearer distinction between driver options and reduction options.

We can make metrics more of a formalized concept that reductions can subscribe to. I think that would be good. However, what I'd really like to avoid is just yet another boolean in the `workspace` object. There should be some level of encapsulation with it.

Reply:
I like the idea of subscribing; something like a callback function registration that can be called after setup and activate any global functionality (i.e. callbacks create metrics related objects where they are needed etc)

communicating via workspace is temptingly easy :)

Reply:
https://github.com/VowpalWabbit/vowpal_wabbit/pull/4354
Opened this which does exactly the `workspace` fix. @jackgerrits how would you suggest we further encapsulate this? It seems like a more global field than many other members in workspace

Reply:
Just adding another boolean to `workspace` isn't really getting us anywhere in my opinion. I think unless we have a more holistic change we shouldn't change anything and just leave it as is. It takes a lot of work to get something out of the god object I really want to avoid doing it without a really good reason.

As for how we further encapsulate this I don't have a design or direction in mind at the moment.

Reply:
This has been resolved with the `metrics_collector`